basic_ds.go

```diff
@@ -166,6 +166,16 @@ func (d *LogDatastore) Delete(key Key) (err error) {
 	return d.child.Delete(key)
 }
 
+func (d *LogDatastore) DiskUsage() (uint64, error) {
+	var du uint64 = 0
+	var err error
+	if pd, ok := d.child.(PersistentDatastore); ok {
+		du, err = pd.DiskUsage()
+	}
+	log.Printf("%s: DiskUsage: %d\n", d.Name, du)
+	return du, err
+}
```

Comment:
This function should return an error if the child is not a PersistentDatastore.

I would also restructure it to avoid having to declare du and err up front.

Something like
```go
pd, ok := d.child.(PersistentDatastore)
if !ok {
  return 0, errors.New("Unimplemented")
}
du, err := pd.DiskUsage()
if err != nil {
  return 0, err
}
log.Printf("%s: DiskUsage: %d\n", d.Name, du)
return du, nil
```
But use a better error message.

Reply:
I see why a batch operation for example, should return an error if the underlying datastore cannot perform it at all. But DiskUsage? This implies that mixing Persistent and non PersistentDatastores in a wrapper and asking for DiskUsage should probably result in error too (as returning error here would mean doing it everywhere for behavior consistency).

If the datastore is not persistent, it seems natural to say (from a wrapper point of view) that the DiskUsage is 0, while keeping the error result to actually signal for real errors when obtaining real disk usage. In the current approach wrappers never produce errors on DiskUsage(), they just bubble them.

Reply:
If the underlying datastore uses disk storage but doesn't implement this method, then a result of 0 without an error would be misleading.  

Maybe we should just require this method be implemented, or perhaps a more generic Stat() method.

@Stebalien thoughts?

Reply:
If a datastore uses the disk, it should report it. If it doesn't, that's a bug in the datastore. However, I don't want to force every datastore to implement this method (it's not really necessary).

I'd like to go with a more generic stat but I couldn't come up with a concrete, nice solution that'll please everyone so, unless you can think of a way to make that work (that's both efficient and idiomatic) we might as well go with this.

Reply:
Okay, I am unconformable with the idea of unimplementable this method means it doesn't use any disk space, but I do agree that will be easier.

I take another look but I think this LGTM then.